packages/ember-application/lib/system/application-instance.js

```diff
@@ -243,14 +243,8 @@ const ApplicationInstance = EngineInstance.extend({
         // No rendering is needed, and routing has completed, simply return.
         return this;
       } else {
-        return new RSVP.Promise((resolve) => {
-          // Resolve once rendering is completed. `router.handleURL` returns the transition (as a thennable)
-          // which resolves once the transition is completed, but the transition completion only queues up
-          // a scheduled revalidation (into the `render` queue) in the Renderer.
-          //
-          // This uses `run.schedule('afterRender', ....)` to resolve after that rendering has completed.
-          run.schedule('afterRender', null, resolve, this);
-        });
+        // Ensure that the visit promise resolves when all rendering has completed
+        return renderSettled().then(() => this);
```

Comment:
@rwjblue Sorry to bring up something old, but seeing some memory leaks in fastboot related to retaining the Container.  Is it necessary to return `this` in the callback?

What I'm worries about is never reaching a `settled` state or resolving, thus every request that comes in builds a new instance (w/ shoebox data) and the eventually swamps the server's heap.

Reply:
On 3.5 right now

Reply:
![Screenshot 2019-04-12 at 13 58 39](https://user-images.githubusercontent.com/7374640/56066747-98cd1780-5d2d-11e9-9a8d-4a1d3474fe30.png)

some more info.  I'm wondering if we need to "settle" the deferred promise when `shouldRender` is false....So not this block but the above block.

Reply:
I don't really follow. Can you file an issue with a reproduction?

Reply:
@rwjblue - We found it.  Well @Serabe found it so I'll let him explain beyond this short answer! In an adapter, he found that `headers` as a computed was potentially leaking the container thus not allowing V8 to not gc anything.  The problem wasn't necessarily ajax/najax/node-fetch but the way we configured our adapter. 

```js
// adapter
headers: computed(function() {
```

to a getter fixed it.

```js
// adapter
get headers() {
```

We are on 3.5